slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.web3signer.slashingprotection;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.jdbi.v3.core.transaction.TransactionIsolationLevel.READ_UNCOMMITTED;
+import static tech.pegasys.web3signer.slashingprotection.DbLocker.lockForValidator;
+
+import tech.pegasys.web3signer.slashingprotection.DbLocker.LockType;
+import tech.pegasys.web3signer.slashingprotection.dao.LowWatermarkDao;
+import tech.pegasys.web3signer.slashingprotection.dao.SignedAttestationsDao;
+import tech.pegasys.web3signer.slashingprotection.dao.SignedBlocksDao;
+import tech.pegasys.web3signer.slashingprotection.dao.SigningWatermark;
+
+import java.util.Optional;
+
+import org.apache.tuweni.units.bigints.UInt64;
+import org.apache.tuweni.units.bigints.UInt64s;
+import org.jdbi.v3.core.Jdbi;
+
+public class Pruner {
+  private final Jdbi jdbi;
+  private final SignedBlocksDao signedBlocksDao;
+  private final SignedAttestationsDao signedAttestationsDao;
+  private final LowWatermarkDao lowWatermarkDao;
+
+  public Pruner(
+      final Jdbi jdbi,
+      final SignedBlocksDao signedBlocksDao,
+      final SignedAttestationsDao signedAttestationsDao,
+      final LowWatermarkDao lowWatermarkDao) {
+    this.jdbi = jdbi;
+    this.signedBlocksDao = signedBlocksDao;
+    this.signedAttestationsDao = signedAttestationsDao;
+    this.lowWatermarkDao = lowWatermarkDao;
+  }
+
+  public void pruneForValidator(
+      final int validatorId, final long epochsToKeep, final long slotsPerEpoch) {
+    checkArgument(
+        epochsToKeep > 0,
+        "epochsToKeep to keep must be a positive value, but was %s",
+        epochsToKeep);
+    checkArgument(
+        slotsPerEpoch > 0, "slotsPerEpoch must be a positive value, but was %", slotsPerEpoch);
+    final long slotsToKeep = Math.max(epochsToKeep / slotsPerEpoch, 1);
+    pruneBlocks(validatorId, slotsToKeep);
+    pruneAttestations(validatorId, epochsToKeep);
+  }
+
+  private void pruneBlocks(final int validatorId, final long slotsToKeep) {
+    final Optional<UInt64> pruningSlot =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.BLOCK, validatorId);
```

Comment:
is there any race condition between importing and Pruning that we hadn't thought of? I suspect its fine, as the watermark can only move upwards.

Reply:
Think it is fine. We advise users not import while the service is running as well so this should also cover this.